File: Scripts/dynamic-mission-test/farps.lua
-- farps.lua — base-role classification + forward FARP network
-- EECH source:
--   ks_dbase.c: AIRBASE (air_force_capacity LARGE, fixed-wing + heli, requires_cap TRUE, requires_
--     barcap FALSE); FARP (air_force_capacity SMALL, HELICOPTERS ONLY).
--   popread.c:1601-1664,1981-2019: airfields with fixed-wing runway routes are AIRBASEs, the rest are
--     helicopter sites; FARP keysites are placed at forward FARP objects. EECH warzones field only a
--     handful of fixed-wing airbases plus a dense field of heli-only bases and forward FARPs.
--
-- Two layers, both EECH-faithful:
--  1) CLASSIFY the DCS airfields (PREDICTABLE structure): a fixed number per side (the REAR fields)
--     run as fixed-wing AIRBASEs; the rest run as heli-only FOBs ("FARP only behaviour").
--  2) Place a FORWARD FARP network (RANDOMNESS/variety): extra heli-only forward bases ahead of the
--     line, jittered in position and count so each campaign's rotary staging differs.
-- Enforcement is via the supply ledger: airbase = fixed-wing + heli; FOB/FARP = helicopters only. So
-- fixed-wing (OCA / keysite strike / SEAD) launches ONLY from the few airbases, while the rotary war
-- (CAS/BAI/insertion) flies from the whole FOB + FARP network — an EECH helicopter-primary front.

local cs     = require("campaign_state")
local config = require("config")
local S  = cs.S
local M  = {}

-- EECH warzones field only "a couple" of fixed-wing airbases per side. Cap (config theatre
-- fixed_wing_per_side), and never more than a third of a side's fields, so the majority stay heli-only
-- FOBs even on a small scoped theatre.
local FIXED_WING_PER_SIDE = config.C.theatre.fixed_wing_per_side
local function fixed_wing_count(side_field_count)
    return math.max(1, math.min(FIXED_WING_PER_SIDE, math.floor(side_field_count / 3)))
end
local FRONT_DIST          = config.C.theatre.farp_front_dist   -- m; a base with an enemy within this is "frontline" (config theatre)
local FARP_FWD_MIN        = 18000    -- m: forward-FARP distance ahead of the parent field (min)
local FARP_FWD_MAX        = 40000    -- m: ... (max) — randomised per FARP
local FARP_LATERAL        = 22000    -- m: max random lateral jitter along the front
local FARP_CHANCE         = 70       -- % chance a given frontline field gets a forward FARP (variety)

local COUNTRY = config.C.countries

-- Nearest enemy base to `base` → (name, distance). math.huge if the enemy has none.
local function nearest_enemy(base, side)
    local bp = S.base_pos[base]
    local enemy = cs.ENEMY[side]
    if not bp or not enemy then return nil, math.huge end
    local best, bestd = nil, math.huge
    for name, owner in pairs(S.base_owner) do
        if owner == enemy and S.base_pos[name] then
            local ep = S.base_pos[name]
            local d = cs.dist2d(bp.x, bp.z, ep.x, ep.z)
            if d < bestd then bestd = d; best = name end
        end
    end
    return best, bestd
end

-- ── Dormant-FARP activation (Item 1 — eech keysite.c:507-568 initialise_keysite_farp_enable) ────────
-- EECH FARP keysites are created in_use=FALSE (ks_creat.c:157) and only ACTIVATE when their SECTOR's
-- side matches their own force side (keysite.c:541-547) — i.e. as the friendly front reaches them.
-- Port proxy on bases-as-sectors: the FARP's "sector side" is the owner of the nearest NON-FARP base
-- to it; the FARP is friendly-sector iff that owner == the FARP's own side. Activation LATCHES (EECH
-- only clears in_use on destroy, ks_dstry.c:325). While dormant a FARP is excluded from launch
-- candidacy (task_board.owned_bases ≈ nearest-keysite search keysite.c:369) and the win usable-airbase
-- census (win_condition ≈ fc_msgs.c:234 ALIVE && IN_USE) via cs.base_is_active.
--
-- WATCH ITEM (F5, documented proxy limit): an AUTHOR-PLACED friendly FARP whose nearest NON-FARP base
-- happens to be enemy-owned reads as "sector not friendly" and starts DORMANT under this proxy — even
-- though the author clearly intends it live (EECH's real sector grid is far finer than
-- bases-as-sectors, so its FARP could still sit in a friendly sector there). It self-heals as the
-- front advances past it; an author who wants such deep-forward FARPs active from boot sets
-- theatre.farp_activation = false (all FARPs active — the documented override).
local function sector_side_friendly(fname)
    local fpos  = S.base_pos[fname]
    local fside = S.base_owner[fname]
    if not fpos or not fside then return false end
    local best_owner, best_d = nil, math.huge
    for name, owner in pairs(S.base_owner) do
        if name ~= fname and (S.base_kind or {})[name] ~= "farp" and S.base_pos[name] then
            local bp = S.base_pos[name]
            local d = cs.dist2d(fpos.x, fpos.z, bp.x, bp.z)
            if d < best_d then best_d = d; best_owner = owner end
        end
    end
    return best_owner == fside
end

-- Seed S.farp_active for every FARP at boot. Gate OFF → all FARPs active (authored convenience).
-- Gate ON (default, EECH-faithful) → active only if the FARP's sector is already friendly; the rest
-- stay dormant until M.update latches them.
function M.seed_activation(log_fn)
    log_fn = log_fn or function() end
    S.farp_active = S.farp_active or {}
    local gate = config.C.theatre.farp_activation
    local n_active, n_dormant = 0, 0
    for name, kind in pairs(S.base_kind or {}) do
        if kind == "farp" then
            local active = (not gate) or sector_side_friendly(name)
            S.farp_active[name] = active or nil
            if active then n_active = n_active + 1 else n_dormant = n_dormant + 1 end
        end
    end
    log_fn(string.format("FARP activation (gate=%s): %d active, %d dormant (dormant activate as the front nears)",
        tostring(gate), n_active, n_dormant))
    cs.dbg("farps", "seed_activation gate=%s: %d active, %d dormant", tostring(gate), n_active, n_dormant)
end

-- Periodic latch: activate any dormant FARP whose sector has become friendly (front advanced). This is
-- the "dormant-FARP activation as the front moves" — EECH re-runs initialise_keysite_farp_enable at
-- campaign setup; the port re-evaluates the same sector-side predicate as ownership shifts.
function M.update(log_fn)
    log_fn = log_fn or function() end
    S.farp_active = S.farp_active or {}
    if not config.C.theatre.farp_activation then return end   -- gate off → all already active
    for name, kind in pairs(S.base_kind or {}) do
        if kind == "farp" and not S.farp_active[name] and sector_side_friendly(name) then
            S.farp_active[name] = true
            log_fn(string.format("FARP %s ACTIVATED — sector now friendly (front reached)", name))
            cs.dbg("farps", "FARP %s ACTIVATED (sector friendly, keysite.c:541 proxy)", name)
        end
    end
end

-- Scheduler (registered once from game_loop.start): re-evaluate FARP activation every 120 s.
function M.schedule(log_fn)
    log_fn = log_fn or function() end
    local my_gen = _DMT_GEN
    local PERIOD = 120
    cs.dbg("farps", "activation scheduler REGISTERED period=%ds", PERIOD)
    timer.scheduleFunction(function(_, t)
        if _DMT_GEN ~= my_gen then return nil end
        pcall(M.update, log_fn)
        return t + PERIOD
    end, nil, timer.getTime() + PERIOD)
end

-- Called from game_loop init AFTER base discovery, BEFORE supply.init (the ledger reads base_kind).
function M.init(log_fn)
    log_fn = log_fn or function() end
    S.base_kind = S.base_kind or {}
    pcall(function() math.randomseed(math.floor((timer.getTime() or 0) * 1000) + cs.GENERATION * 7919) end)

    -- ── AUTHORED theatre (zone-based) — REGISTER, do not spawn ────────────────────
    -- When the bases were authored by ME keysite zones, keysite.lua already set base_kind
    -- ("airbase"/"farp") from the zones. The author PLACES the FARP static in the Mission Editor
    -- (it auto-registers as a HELIPAD airbase); the campaign SPAWNS NOTHING here. We only confirm the
    -- farp roles. The supply-ledger contract (base_kind "airbase" = fixed-wing+heli; "farp" =
    -- heli-only) is already set by keysite.lua and preserved.
    if require("zones").has_keysite_zones() then
        local nfarp = 0
        for _, kind in pairs(S.base_kind) do
            if kind == "farp" then nfarp = nfarp + 1 end
        end
        log_fn(string.format("base roles from zones: %d farp bases confirmed (author-placed FARP "
            .. "statics; NO spawning) — airbase/farp roles set by keysite zones", nfarp))
        cs.dbg("farps", "init (zone-authored): %d farp bases confirmed, no spawning", nfarp)
        M.seed_activation(log_fn)   -- Item 1: dormant-FARP activation seeding (zone-authored FARPs)
        return
    end

    -- ── Layer 1: classify airfields (predictable) ────────────────────────────────
    local nfw, nfob = 0, 0
    for _, side in ipairs({ coalition.side.BLUE, coalition.side.RED }) do
        local fields = {}
        for name, owner in pairs(S.base_owner) do
            if owner == side then
                local _, d = nearest_enemy(name, side)
                fields[#fields + 1] = { name = name, d = d }
            end
        end
        -- Rear-most fields (largest distance to the enemy) are the fixed-wing AIRBASEs; the forward
        -- fields become heli-only FOBs. The airbase count scales down on a small theatre so the
        -- majority always stay heli (never more than a third of a side's fields, capped at 2).
        table.sort(fields, function(a, b) return a.d > b.d end)
        local nfw_side = fixed_wing_count(#fields)
        for i, f in ipairs(fields) do
            if i <= nfw_side then S.base_kind[f.name] = "airbase"; nfw = nfw + 1
            else S.base_kind[f.name] = "fob"; nfob = nfob + 1 end
        end
    end

    -- ── Layer 2: forward FARP network (randomised) ───────────────────────────────
    local airfields = {}
    for name, owner in pairs(S.base_owner) do
        if S.base_kind[name] == "airbase" or S.base_kind[name] == "fob" then
            airfields[#airfields + 1] = { name = name, owner = owner }
        end
    end
    local nfarp = 0
    for _, ab in ipairs(airfields) do
        local enemy_base, d = nearest_enemy(ab.name, ab.owner)
        if enemy_base and d < FRONT_DIST and math.random(100) <= FARP_CHANCE then
            local bp, ep = S.base_pos[ab.name], S.base_pos[enemy_base]
            local dx, dz = ep.x - bp.x, ep.z - bp.z
            local len = math.max(math.sqrt(dx * dx + dz * dz), 1)
            local ux, uz = dx / len, dz / len          -- unit vector toward the enemy
            local px, pz = -uz, ux                       -- perpendicular (lateral) unit vector
            local fwd = FARP_FWD_MIN + math.random() * (FARP_FWD_MAX - FARP_FWD_MIN)
            local lat = (math.random() * 2 - 1) * FARP_LATERAL
            -- Snap the forward-FARP placement off water (a jittered point 18-40 km ahead of a coastal
            -- field can land in the sea); fallback = the parent field centre `bp` (always land).
            local fp = cs.snap_land(bp.x + ux * fwd + px * lat, bp.z + uz * fwd + pz * lat, bp.x, bp.z)
            local fx = fp.x
            local fz = fp.z
            local fname = "FARP-" .. ab.name:sub(1, 12)

            S.base_owner[fname]  = ab.owner
            S.base_pos[fname]    = { x = fx, z = fz }
            S.base_health[fname] = 1.0
            S.base_kind[fname]   = "farp"

            -- Idempotent: FARP statics survive reset.nuke (it destroys groups, not statics).
            if not Airbase.getByName(fname) then
                local fs = config.C.statics.farp   -- config statics.farp = {type, shape_name, category}
                coalition.addStaticObject(COUNTRY[ab.owner], {
                    heading = 0, type = fs.type, shape_name = fs.shape_name, category = fs.category,
                    name = fname, x = fx, y = fz, dead = false,
                })
            end
            nfarp = nfarp + 1
        end
    end

    log_fn(string.format("base roles: %d fixed-wing AIRBASEs (rear) + %d heli-only FOBs + %d forward "
        .. "FARPs — fixed-wing flies only from airbases; the rotary front war flies from FOBs + FARPs",
        nfw, nfob, nfarp))
    cs.dbg("farps", "init (auto): %d airbases, %d heli-only FOBs, %d forward FARPs placed", nfw, nfob, nfarp)
    M.seed_activation(log_fn)   -- Item 1: dormant-FARP activation seeding (auto-theatre forward FARPs)
end

-- NOTE: M.is_heli_only was DELETED in Cluster H (zero callers) — heli-only enforcement lives in the
-- ledger seeding (supply.lua: FOB/FARP bases get no fixed-wing striker/escort/recon roles), which is
-- what actually confines fixed-wing to the airbases.

return M

File: packages/ee-mission/src/farps.ts
/** @noSelfInFile */
/*
-- farps.lua — base-role classification + forward FARP network
-- EECH source:
--   ks_dbase.c: AIRBASE (air_force_capacity LARGE, fixed-wing + heli, requires_cap TRUE, requires_
--     barcap FALSE); FARP (air_force_capacity SMALL, HELICOPTERS ONLY).
--   popread.c:1601-1664,1981-2019: airfields with fixed-wing runway routes are AIRBASEs, the rest are
--     helicopter sites; FARP keysites are placed at forward FARP objects. EECH warzones field only a
--     handful of fixed-wing airbases plus a dense field of heli-only bases and forward FARPs.
--
-- Two layers, both EECH-faithful:
--  1) CLASSIFY the DCS airfields (PREDICTABLE structure): a fixed number per side (the REAR fields)
--     run as fixed-wing AIRBASEs; the rest run as heli-only FOBs ("FARP only behaviour").
--  2) Place a FORWARD FARP network (RANDOMNESS/variety): extra heli-only forward bases ahead of the
--     line, jittered in position and count so each campaign's rotary staging differs.
-- Enforcement is via the supply ledger: airbase = fixed-wing + heli; FOB/FARP = helicopters only. So
-- fixed-wing (OCA / keysite strike / SEAD) launches ONLY from the few airbases, while the rotary war
-- (CAS/BAI/insertion) flies from the whole FOB + FARP network — an EECH helicopter-primary front.
*/

import * as cs from "./campaign_state";
import type { Side } from "./campaign_types";
import * as config from "./config";
import { BLUE, RED } from "./sides";
import * as zones from "./zones";

type LogFunction = (this: void, message: string) => void;
interface RatedField {
	name: string;
	distance: number;
}
interface FieldOwner {
	name: string;
	owner: Side;
}

const S = cs.S;
const SIDES: Side[] = [BLUE, RED];
const FIXED_WING_PER_SIDE = config.C.theatre.fixed_wing_per_side;
const FRONT_DIST = config.C.theatre.farp_front_dist;
// DCS adapter placement envelope in metres; EECH uses authored FARP objects rather than offsets.
const FARP_FORWARD_MIN_METRES = 18000;
const FARP_FORWARD_MAX_METRES = 40000;
const FARP_LATERAL_JITTER_METRES = 22000;
const FARP_PLACEMENT_CHANCE_PERCENT = 70;
const RANDOM_PERCENT_MAX = 100;
// DCS adapter cadence; EECH activation is entity-driven, so poll every two minutes.
const ACTIVATION_POLL_SECONDS = 120;
// Deterministic seed mixing constants for milliseconds and generation separation.
const MILLISECONDS_PER_SECOND = 1000;
const GENERATION_SEED_PRIME = 7919;
// Preserve at least one fixed-wing field and no more than one third of available fields.
const FIXED_WING_FIELD_DIVISOR = 3;
const COUNTRY = config.C.countries;
// Mission Editor-authored stock FARP defaults, verified from a DCS 2.9 mission:
// type=FARP, shape_name=FARPS, 127.5 MHz AM and callsign id 1 (London).
const STOCK_FARP_FREQUENCY_MHZ = "127.5";
const STOCK_FARP_AM_MODULATION = 0;
const STOCK_FARP_CALLSIGN_ID = 1;

function ensurePhysicalFarp(
	name: string,
	owner: Side,
	position: { x: number; z: number },
): boolean {
	const [airbaseRead, airbase] = pcall(() => Airbase.getByName(name));
	if (airbaseRead && airbase !== undefined) return false;
	const [staticRead, staticObject] = pcall(() => StaticObject.getByName(name));
	if (staticRead && staticObject?.isExist()) return false;

	const farpStatic = config.C.statics.farp;
	coalition.addStaticObject(COUNTRY[owner], {
		heading: 0,
		type: farpStatic.type,
		shape_name: farpStatic.shape_name,
		category: farpStatic.category,
		name,
		x: position.x,
		y: position.z,
		dead: false,
		heliport_frequency: STOCK_FARP_FREQUENCY_MHZ,
		heliport_modulation: STOCK_FARP_AM_MODULATION,
		heliport_callsign_id: STOCK_FARP_CALLSIGN_ID,
	});
	return true;
}

function fixedWingCount(sideFieldCount: number): number {
	return math.max(
		1,
		math.min(
			FIXED_WING_PER_SIDE,
			math.floor(sideFieldCount / FIXED_WING_FIELD_DIVISOR),
		),
	);
}

function nearestEnemy(
	base: string,
	side: Side,
): LuaMultiReturn<[string | undefined, number]> {
	const basePos = S.base_pos[base];
	const enemy = cs.ENEMY[side];
	if (basePos === undefined || enemy === undefined)
		return $multi(undefined, math.huge);
	let best: string | undefined;
	let bestDistance = math.huge;
	for (const [name, owner] of pairs(S.base_owner)) {
		const enemyPos = S.base_pos[name];
		if (owner === enemy && enemyPos !== undefined) {
			const distance = cs.dist2d(basePos.x, basePos.z, enemyPos.x, enemyPos.z);
			if (distance < bestDistance) {
				bestDistance = distance;
				best = name;
			}
		}
	}
	return $multi(best, bestDistance);
}

// keysite.c:507-568 initialise_keysite_farp_enable: FARPs start out of use and activate only when
// their sector side matches their force side. Bases are the port's sector proxy; activation latches.
function sectorSideFriendly(farpName: string): boolean {
	const farpPos = S.base_pos[farpName];
	const farpSide = S.base_owner[farpName];
	if (farpPos === undefined || farpSide === undefined) return false;
	let bestOwner: Side | undefined;
	let bestDistance = math.huge;
	for (const [name, owner] of pairs(S.base_owner)) {
		const basePos = S.base_pos[name];
		if (
			name !== farpName &&
			S.base_kind[name] !== "farp" &&
			basePos !== undefined
		) {
			const distance = cs.dist2d(farpPos.x, farpPos.z, basePos.x, basePos.z);
			if (distance < bestDistance) {
				bestDistance = distance;
				bestOwner = owner;
			}
		}
	}
	return bestOwner === farpSide;
}

export function seed_activation(logFn: LogFunction = () => undefined): void {
	S.farp_active ??= {};
	const gate = config.C.theatre.farp_activation;
	let activeCount = 0;
	let dormantCount = 0;
	for (const [name, kind] of pairs(S.base_kind)) {
		if (kind === "farp") {
			const active = !gate || sectorSideFriendly(name);
			if (active) {
				S.farp_active[name] = true;
				activeCount++;
			} else {
				delete S.farp_active[name];
				dormantCount++;
			}
		}
	}
	logFn(
		string.format(
			"FARP activation (gate=%s): %d active, %d dormant (dormant activate as the front nears)",
			tostring(gate),
			activeCount,
			dormantCount,
		),
	);
	cs.dbg(
		"farps",
		"seed_activation gate=%s: %d active, %d dormant",
		tostring(gate),
		activeCount,
		dormantCount,
	);
}

export function update(logFn: LogFunction = () => undefined): void {
	S.farp_active ??= {};
	if (!config.C.theatre.farp_activation) return;
	for (const [name, kind] of pairs(S.base_kind)) {
		if (kind === "farp" && !S.farp_active[name] && sectorSideFriendly(name)) {
			S.farp_active[name] = true;
			logFn(
				string.format(
					"FARP %s ACTIVATED — sector now friendly (front reached)",
					name,
				),
			);
			cs.dbg(
				"farps",
				"FARP %s ACTIVATED (sector friendly, keysite.c:541 proxy)",
				name,
			);
		}
	}
}

export function schedule(logFn: LogFunction = () => undefined): void {
	const generation = _DMT_GEN;
	const period = ACTIVATION_POLL_SECONDS;
	cs.dbg("farps", "activation scheduler REGISTERED period=%ds", period);
	timer.scheduleFunction(
		(_argument: undefined, time: number) => {
			if (_DMT_GEN !== generation) return undefined;
			pcall(() => update(logFn));
			return time + period;
		},
		undefined,
		timer.getTime() + period,
	);
}

export function init(logFn: LogFunction = () => undefined): void {
	pcall(() =>
		math.randomseed(
			math.floor(timer.getTime() * MILLISECONDS_PER_SECOND) +
				cs.GENERATION * GENERATION_SEED_PRIME,
		),
	);

	// Authored theatre: generated missions carry FARP trigger zones, not Mission Editor heliport
	// objects. Materialise one physical DCS FARP at each logical base so it is visible and usable.
	if (zones.has_keysite_zones()) {
		let farpCount = 0;
		let spawnedCount = 0;
		for (const [name, kind] of pairs(S.base_kind)) {
			if (kind !== "farp") continue;
			farpCount++;
			const owner = S.base_owner[name];
			const position = S.base_pos[name];
			if (
				(owner === BLUE || owner === RED) &&
				position !== undefined &&
				ensurePhysicalFarp(name, owner, position)
			)
				spawnedCount++;
		}
		logFn(
			string.format(
				"base roles from zones: %d farp bases confirmed, %d physical FARP statics queued — airbase/farp roles set by keysite zones",
				farpCount,
				spawnedCount,
			),
		);
		cs.dbg(
			"farps",
			"init (zone-authored): %d farp bases confirmed, %d physical statics queued",
			farpCount,
			spawnedCount,
		);
		seed_activation(logFn);
		return;
	}

	let fixedWingCountTotal = 0;
	let fobCount = 0;
	for (const side of SIDES) {
		const fields: RatedField[] = [];
		for (const [name, owner] of pairs(S.base_owner)) {
			if (owner === side) {
				const [, distance] = nearestEnemy(name, side);
				fields.push({ name, distance });
			}
		}
		fields.sort((a, b) => b.distance - a.distance);
		const fixedWingForSide = fixedWingCount(fields.length);
		for (let index = 0; index < fields.length; index++) {
			if (index < fixedWingForSide) {
				S.base_kind[fields[index].name] = "airbase";
				fixedWingCountTotal++;
			} else {
				S.base_kind[fields[index].name] = "fob";
				fobCount++;
			}
		}
	}

	const airfields: FieldOwner[] = [];
	for (const [name, owner] of pairs(S.base_owner)) {
		if (S.base_kind[name] === "airbase" || S.base_kind[name] === "fob") {
			airfields.push({ name, owner });
		}
	}
	let farpCount = 0;
	for (const airfield of airfields) {
		const [enemyBase, distance] = nearestEnemy(airfield.name, airfield.owner);
		if (
			enemyBase !== undefined &&
			distance < FRONT_DIST &&
			math.random(RANDOM_PERCENT_MAX) <= FARP_PLACEMENT_CHANCE_PERCENT
		) {
			const basePos = S.base_pos[airfield.name];
			const enemyPos = S.base_pos[enemyBase];
			const dx = enemyPos.x - basePos.x;
			const dz = enemyPos.z - basePos.z;
			const length = math.max(math.sqrt(dx * dx + dz * dz), 1);
			const unitX = dx / length;
			const unitZ = dz / length;
			const perpendicularX = -unitZ;
			const perpendicularZ = unitX;
			const forward =
				FARP_FORWARD_MIN_METRES +
				math.random() * (FARP_FORWARD_MAX_METRES - FARP_FORWARD_MIN_METRES);
			const lateral = (math.random() * 2 - 1) * FARP_LATERAL_JITTER_METRES;
			const farpPos = cs.snap_land(
				basePos.x + unitX * forward + perpendicularX * lateral,
				basePos.z + unitZ * forward + perpendicularZ * lateral,
				basePos.x,
				basePos.z,
			);
			const farpName = `FARP-${string.sub(airfield.name, 1, 12)}`;
			S.base_owner[farpName] = airfield.owner;
			S.base_pos[farpName] = { x: farpPos.x, z: farpPos.z };
			S.base_health[farpName] = 1;
			S.base_kind[farpName] = "farp";

			ensurePhysicalFarp(farpName, airfield.owner, farpPos);
			farpCount++;
		}
	}

	logFn(
		string.format(
			"base roles: %d fixed-wing AIRBASEs (rear) + %d heli-only FOBs + %d forward FARPs — fixed-wing flies only from airbases; the rotary front war flies from FOBs + FARPs",
			fixedWingCountTotal,
			fobCount,
			farpCount,
		),
	);
	cs.dbg(
		"farps",
		"init (auto): %d airbases, %d heli-only FOBs, %d forward FARPs placed",
		fixedWingCountTotal,
		fobCount,
		farpCount,
	);
	seed_activation(logFn);
}
